src/webgpu/shader/execution/expression/call/builtin/textureDimensions.spec.ts

```diff
@@ -0,0 +1,497 @@
+    case '1d': {
+      const w = bw * 8;
+      return { size: [w], expected: [w >>> mip] };
+    }
+    case '2d': {
+      const w = bw * 4;
+      const h = bh * 8;
+      return { size: [w, h], expected: [w >>> mip, h >>> mip] };
+    }
+    case '2d-array': {
+      const w = bw * 8;
+      const h = bh * 4;
+      return { size: [w, h, 4], expected: [w >>> mip, h >>> mip] };
+    }
+    case '3d': {
+      const w = bw * 8;
+      const h = bh * 4;
+      const d = 16;
+      return {
+        size: [w, h, d],
+        expected: [w >>> mip, h >>> mip, d >>> mip],
+      };
+    }
+    case 'cube': {
+      const l = bw * bh * 4;
+      return {
+        size: [l, l, 6],
+        expected: [l >>> mip, l >>> mip],
+      };
+    }
+    case 'cube-array': {
+      const l = bw * bh * 4;
+      return {
+        size: [l, l, 12],
+        expected: [l >>> mip, l >>> mip],
+      };
+    }
+  }
+}
+
+/**
+ * Builds a shader module with the texture view bound to the WGSL texture with the given WGSL type,
+ * which calls textureDimensions(), assigning the result to an output buffer.
+ * This shader is executed with a compute shader, and the output buffer is compared to
+ * `values.expected`.
+ */
+function run(
+  t: GPUTest,
+  view: GPUTextureView,
+  textureType: string,
+  levelArg: number | undefined,
+  values: TestValues
+) {
+  const outputType = values.expected.length > 1 ? `vec${values.expected.length}u` : 'u32';
+  const wgsl = `
+@group(0) @binding(0) var texture : ${textureType};
+@group(0) @binding(1) var<storage, read_write> output : ${outputType};
+
+@compute @workgroup_size(1)
+fn main() {
+output = ${
+    levelArg !== undefined
+      ? `textureDimensions(texture, ${levelArg})`
+      : 'textureDimensions(texture)'
+  };
+}
+`;
+  const module = t.device.createShaderModule({
+    code: wgsl,
+  });
+  const pipeline = t.device.createComputePipeline({
+    compute: { module },
+    layout: 'auto',
+  });
+  const outputBuffer = t.device.createBuffer({
+    size: 32,
+    usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.STORAGE,
+  });
+  const bindgroup = t.device.createBindGroup({
+    layout: pipeline.getBindGroupLayout(0),
+    entries: [
+      { binding: 0, resource: view },
+      { binding: 1, resource: { buffer: outputBuffer } },
+    ],
+  });
+  const encoder = t.device.createCommandEncoder();
+  const pass = encoder.beginComputePass();
+  pass.setPipeline(pipeline);
+  pass.setBindGroup(0, bindgroup);
+  pass.dispatchWorkgroups(1);
+  pass.end();
+  t.device.queue.submit([encoder.finish()]);
+
+  t.expectGPUBufferValuesEqual(outputBuffer, new Uint32Array(values.expected));
+}
+
+/** @returns true if the GPUTextureViewDimension is valid for a storage texture */
+function dimensionsValidForStorage(dimensions: GPUTextureViewDimension) {
+  switch (dimensions) {
+    case '1d':
+    case '2d':
+    case '2d-array':
+    case '3d':
+      return true;
+    default:
+      return false;
+  }
+}
+
+g.test('sampled')
```

Comment:
nit:  A mulitsampled texture is not a sampled texture.  (You don't use a sampler with it.)
Yes, it's confusing.  But worth a comment here to say that this test covers both sampled and multisampled textures.

Reply:
Done